### PR TITLE
[SPARK-15735] Allow specifying min time to run in microbenchmarks

### PR DESCRIPTION
**Pull request by ericl**
## What changes were proposed in this pull request?

This makes microbenchmarks run for at least 2 seconds by default, to allow some time for jit compilation to kick in.
## How was this patch tested?

Tested manually with existing microbenchmarks. This change is backwards compatible in that existing microbenchmarks which specified numIters per-case will still run exactly that number of iterations. Microbenchmarks which previously overrode defaultNumIters now override minNumIters.

cc @hvanhovell

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala [hunk 1]
```diff
@@ -97,6 +111,39 @@ private[spark] class Benchmark(
...
+    System.gc()  // ensures garbage from previous cases don't impact this one
+    val minIters = if (overrideNumIters != 0) overrideNumIters else minNumIters
+    val minDuration = if (overrideNumIters != 0) 0.seconds.fromNow else minTime.fromNow
+    val runTimes = ArrayBuffer[Long]()
+    var i = 0
+    while (i < minIters || !minDuration.isOverdue) {
+      val timer = new Benchmark.Timer(i)
+      f(timer)
+      val runTime = timer.totalTime()
+      if (i > 0) {
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: This is a left-over from the previous version. Any idea why it was added? One warm-up period?

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala [hunk 2]
```diff
@@ -97,6 +111,39 @@ private[spark] class Benchmark(
...
+        runTimes += runTime
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: It is quite likely that we will also add unoptimized results here. Is that a problem? Or are we only interested in the `best`  runtime?

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala (hunk 1 above)
Comment (ericl) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: I believe so. I would not be opposed to removing it.

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala (hunk 2 above)
Comment (ericl) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: We could extend the warmup period from 1 iteration to a specified time, not sure if it's worth it though.

**Comment by ericl** [comment]
In the interest of more reproducible results, added a warmup time and option to output to a file. cc @JoshRosen

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala [hunk 3]
```diff
@@ -33,18 +38,37 @@ import org.apache.commons.lang3.SystemUtils
...
+ * @param outputPerIteration if true, the timing for each run will be printed to stdout.
+ * @param output optional output stream to write benchmark results to
  */
 private[spark] class Benchmark(
     name: String,
     valuesPerIteration: Long,
-    defaultNumIters: Int = 5,
-    outputPerIteration: Boolean = false) {
+    minNumIters: Int = 2,
+    warmupTime: FiniteDuration = 2.seconds,
```
Comment (kiszk) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: I like to add a warmup period, but is it better to 2 second as default. It depends on machine, program (we can specify a value explicitly), and others.
Are there any other thoughts?

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala [hunk 4]
```diff
@@ -33,18 +38,37 @@ import org.apache.commons.lang3.SystemUtils
...
+    minTime: FiniteDuration = 2.seconds,
```
Comment (kiszk) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: ditto.

File: core/src/main/scala/org/apache/spark/util/Benchmark.scala (hunk 3 above)
Comment (ericl) [inline core/src/main/scala/org/apache/spark/util/Benchmark.scala]: I suppose the relevant tradeoff here is accuracy vs time added to benchmark. I'd like to pick a value that is a reasonable default for most benchmarks - probably 2-5 seconds to warm up the jit reasonably depending on the program. Those with specialized needs can override it as needed.

**Comment by JoshRosen** [comment]
This seems good to me. @hvanhovell, did you have any comments / do you want to do the final sign-off?

**Comment by hvanhovell** [comment]
LGTM

**Comment by hvanhovell** [comment]
Merging to master/2.0

**Pull request closed without merging**